Core/Core/Grades/Model/GradeListInteractor.swift

```diff
@@ -310,69 +310,126 @@ public final class GradeListInteractorLive: GradeListInteractor {
+                gradeEnrollment: gradeEnrollment,
+                course: course
+            ))
+            .eraseToAnyPublisher()
+        } else {
+            return Just(getGradeForShowQuantitativeData(
+                baseOnGradedAssignments: baseOnGradedAssignments,
+                courseEnrollment: courseEnrollment,
+                gradeEnrollment: gradeEnrollment
+            ))
+            .eraseToAnyPublisher()
+        }
+    }
+
+    private func getGradeForHideQuantitativeData(
+        baseOnGradedAssignments: Bool,
+        courseEnrollment: Enrollment?,
+        gradeEnrollment: Enrollment?,
+        course: Course
+    ) -> String? {
+        if let gradingPeriodID {
+            return getGradeForGradingPeriod(gradingPeriodID: gradingPeriodID)
+        } else {
+            return getGradeForNoGradingPeriod()
+        }
+
+        func getGradeForGradingPeriod(gradingPeriodID: String) -> String? {
+            let letterGrade = baseOnGradedAssignments
+                ? gradeEnrollment?.currentGrade(gradingPeriodID: gradingPeriodID)
+                : gradeEnrollment?.finalGrade(gradingPeriodID: gradingPeriodID)
+
+            if let letterGrade {
+                return letterGrade
             } else {
-                if courseEnrollment?.multipleGradingPeriodsEnabled == true, courseEnrollment?.totalsForAllGradingPeriodsOption == false {
-                    return Just(nil).eraseToAnyPublisher()
-                } else if let letterGrade = courseEnrollment?.computedCurrentGrade ?? courseEnrollment?.computedFinalGrade ?? courseEnrollment?.computedCurrentLetterGrade {
-                    return Just(letterGrade).eraseToAnyPublisher()
-                } else {
-                    return Just(courseEnrollment?.convertedLetterGrade(
-                        gradingPeriodID: nil,
-                        gradingScheme: course.gradingScheme
-                    )
-                    ).eraseToAnyPublisher()
-                }
+                return gradeEnrollment?.convertedLetterGrade(
+                    gradingPeriodID: gradingPeriodID,
+                    gradingScheme: course.gradingScheme
+                )
+            }
+        }
+
+        func getGradeForNoGradingPeriod() -> String? {
+            let letterGrade = (
+                baseOnGradedAssignments
+                ? courseEnrollment?.computedCurrentGrade
+                : courseEnrollment?.computedFinalGrade
+            ) ?? courseEnrollment?.computedCurrentLetterGrade
+
+            if courseEnrollment?.multipleGradingPeriodsEnabled == true,
+               courseEnrollment?.totalsForAllGradingPeriodsOption == false {
+                return nil
+            } else if let letterGrade {
+                return letterGrade
+            } else {
+                return courseEnrollment?.convertedLetterGrade(
+                    gradingPeriodID: nil,
+                    gradingScheme: course.gradingScheme
+                )
```

Comment:
Can we simplify this function to be reduce using optional with operations, such way can be hard to diagnose given those optional hide uncertain condition.

So, I'd suggest the function to start with `courseEnrollment` guard. So, to make to clear what's expected at this state.